torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -755,6 +757,80 @@ XLAGraphExecutor::ExecuteComputationWithBarrier(
   return placeholders;
 }
 
+std::vector<torch::lazy::BackendDataPtr> XLAGraphExecutor::ExecuteStablehlo(
+    std::string bytecode, const std::vector<at::IValue>& graph_inputs,
+    const torch::lazy::BackendDevice& device) {
+  // Convert StableHLO to HLO for XLA compilation.
+  // TODO(lsy323): Pass StableHLO to PjrtComputationClient for compilation
+  // after StableHLO compilation API is added in ComputationClient.
+  mlir::MLIRContext context;
+  mlir::OwningOpRef<mlir::ModuleOp> module =
+      mlir::stablehlo::deserializePortableArtifact(bytecode, &context);
+  mlir::ModuleOp mlir_module = *module;
+  xla::HloProto hlo_proto;
+  runtime::ConvertStableHloToHlo(&mlir_module, &context, &hlo_proto);
+  xla::HloModuleProto* hlo_module_proto = hlo_proto.mutable_hlo_module();
+  xla::XlaComputation computation(*hlo_module_proto);
+
+  // Get program output shape.
+  // TODO(lsy323): Get shape info from MLIR Module.
+  xla::ProgramShape program_shape = ConsumeValue(computation.GetProgramShape());
+  xla::Shape shape = MakeShapeWithDeviceLayout(
+      program_shape.result(), static_cast<XlaDeviceType>(device.type()));
+  std::vector<xla::Shape> flatten_shapes;
+  if (shape.IsTuple()) {
+    // The output shape is a Tuple if graph has multiple outputs.
+    flatten_shapes = std::move(shape.tuple_shapes());
+  } else {
+    flatten_shapes.push_back(std::move(shape));
+  }
+
+  std::vector<runtime::ComputationClient::CompileInstance> instances;
+  instances.push_back({std::move(computation), device.toString(),
+                       runtime::GetComputationClient()->GetCompilationDevices(
+                           device.toString(),
+                           runtime::GetComputationClient()->GetLocalDevices()),
+                       &shape});
+  std::vector<std::shared_ptr<runtime::ComputationClient::Computation>>
+      computations =
+          runtime::GetComputationClient()->Compile(std::move(instances));
+
+  // Placeholder for computation outputs.
+  std::vector<torch::lazy::BackendDataPtr> placeholders;
+  for (const auto& s : flatten_shapes) {
+    torch::lazy::BackendDataPtr handle =
+        WrapXlaData(runtime::GetComputationClient()->CreateDataPlaceholder(
+            device.toString(), std::move(s)));
+    placeholders.push_back(handle);
+  }
```

Comment:
you don't really need placeholders if the computation happens on the same thread. The purpose of placeholder is this function can return and the actual compuations happens in a background thread.

Reply:
Thanks Jack, updated